#**Question 1 : Explain the fundamental differences between DDL, DML, and DQL commands in SQL. Provide one example for each type of command.**

**Answer :**

SQL commands are categorized based on the type of operations they perform on the database.

**1. DDL (Data Definition Language)**

DDL commands are used to define, modify, or delete the structure of database objects such as tables, schemas, and indexes.

Examples of DDL commands:

CREATE

ALTER

DROP

TRUNCATE


**2. DML (Data Manipulation Language)**

DML commands are used to manipulate data stored in database tables.

Examples of DML commands:

INSERT

UPDATE

DELETE


**3. DQL (Data Query Language)**

DQL commands are used to retrieve data from the database.

Example of DQL command:

SELECT


In [ ]:
import sqlite3

# Connect to an in-memory SQLite database
conn = sqlite3.connect(':memory:')
cursor = conn.cursor()

print("Connected to an in-memory SQLite database.")

Connected to an in-memory SQLite database.


### Execute DDL Command (CREATE TABLE)

First, let's create the `Students` table using the DDL `CREATE TABLE` command.

In [ ]:
ddl_command = """
CREATE TABLE Students (
    StudentID INT PRIMARY KEY,
    StudentName VARCHAR(50)
);
"""
cursor.execute(ddl_command)
conn.commit()
print("Table 'Students' created successfully.")

Table 'Students' created successfully.


### Execute DML Command (INSERT INTO)

Next, we'll insert a record into the `Students` table using the DML `INSERT INTO` command.

In [ ]:
dml_command = "INSERT INTO Students VALUES (1, 'Shreya');"
cursor.execute(dml_command)
conn.commit()
print("Data inserted successfully.")

Data inserted successfully.


### Execute DQL Command (SELECT)

Finally, let's retrieve the data from the `Students` table using the DQL `SELECT` command.

In [ ]:
dql_command = "SELECT * FROM Students;"
cursor.execute(dql_command)
results = cursor.fetchall()

print("Query Results:")
for row in results:
    print(row)

# Close the connection (optional for in-memory, but good practice)
conn.close()
print("\nDatabase connection closed.")

Query Results:
(1, 'Shreya')

Database connection closed.


#**Question 2 : What is the purpose of SQL constraints? Name and describe three common types of constraints, providing a simple scenario where each would be useful.**
**Answer :**

SQL constraints are rules applied to table columns to ensure the accuracy, reliability, and integrity of data.

**1. PRIMARY KEY**

Ensures each record is unique and not null.

Used to uniquely identify records.

Scenario:
CustomerID in a Customers table.

**2. NOT NULL**

Ensures a column cannot have NULL values.

Scenario:
CustomerName should never be empty.

**3. UNIQUE**

Ensures all values in a column are unique.

Scenario:
Email addresses in a customer table must be unique.

#**Question 3 : Explain the difference between LIMIT and OFFSET clauses in SQL. How would you use them together to retrieve the third page of results, assuming each page has 10 records?**

**Answer :**

LIMIT specifies the number of records to retrieve.

OFFSET specifies how many records to skip before starting retrieval.

**Third page calculation:**

Records per page = 10

Records to skip = (3 − 1) × 10 = 20



In [ ]:
sql_query = "SELECT * FROM YourTableName LIMIT 10 OFFSET 20;"
print(sql_query)

SELECT * FROM YourTableName LIMIT 10 OFFSET 20;


#**Question 4 : What is a Common Table Expression (CTE) in SQL, and what are its main benefits? Provide a simple SQL example demonstrating its usage.**

**Answer :**

A CTE (Common Table Expression) is a temporary named result set defined using the WITH keyword.

Benefits:

Improves query readability

Simplifies complex queries

Can be reused within the same query


In [ ]:
import sqlite3

# Connect to a new in-memory SQLite database for the CTE example
conn_cte = sqlite3.connect(':memory:')
cursor_cte = conn_cte.cursor()

print("Connected to a new in-memory SQLite database for CTE example.")

Connected to a new in-memory SQLite database for CTE example.


### Create the `Orders` Table and Insert Sample Data

In [ ]:
create_table_sql = """
CREATE TABLE Orders (
    OrderID INT PRIMARY KEY,
    Region VARCHAR(50),
    SalesAmount DECIMAL(10, 2)
);
"""
cursor_cte.execute(create_table_sql)
conn_cte.commit()
print("Table 'Orders' created successfully.")

insert_data_sql = """
INSERT INTO Orders (OrderID, Region, SalesAmount) VALUES
(1, 'North', 50000.00),
(2, 'South', 120000.00),
(3, 'East', 80000.00),
(4, 'West', 150000.00),
(5, 'North', 70000.00),
(6, 'South', 90000.00),
(7, 'East', 110000.00),
(8, 'West', 60000.00);
"""
cursor_cte.execute(insert_data_sql)
conn_cte.commit()
print("Sample data inserted into 'Orders' table.")

Table 'Orders' created successfully.
Sample data inserted into 'Orders' table.


### Execute the CTE Query

In [ ]:
cte_query = """
WITH RegionalSales AS (
    SELECT
        Region,
        SUM(SalesAmount) AS TotalSales
    FROM
        Orders
    GROUP BY
        Region
)
SELECT
    Region,
    TotalSales
FROM
    RegionalSales
WHERE
    TotalSales > 100000;
"""

cursor_cte.execute(cte_query)
cte_results = cursor_cte.fetchall()

print("CTE Query Results (Regions with Total Sales > 100,000):")
for row in cte_results:
    print(row)

# Close the connection
conn_cte.close()
print("\nDatabase connection for CTE example closed.")

CTE Query Results (Regions with Total Sales > 100,000):
('East', 190000)
('North', 120000)
('South', 210000)
('West', 210000)

Database connection for CTE example closed.


#**Question 5 : Describe the concept of SQL Normalization and its primary goals. Briefly explain the first three normal forms (1NF, 2NF, 3NF).**

**Answer :**

Normalization is the process of organizing data to reduce redundancy and improve data integrity.

**Primary Goals:**

Eliminate duplicate data

Ensure logical data storage

Maintain consistency

**1NF (First Normal Form):**

Atomic values

No repeating groups

**2NF (Second Normal Form):**

Must be in 1NF

No partial dependency on composite keys

**3NF (Third Normal Form):**

Must be in 2NF

No transitive dependencies

#**Question 6 :**

#**Create a database named ECommerceDB and perform the following tasks:**

#**1. Create the following tables with appropriate data types and constraints:**

**Categories:**

    CategoryID (INT, PRIMARY KEY)

    CategoryName (VARCHAR(50), NOT NULL, UNIQUE)

**Products:**

    ProductID (INT, PRIMARY KEY)

    ProductName (VARCHAR(100), NOT NULL, UNIQUE)

    CategoryID (INT, FOREIGN KEY → Categories)

    Price (DECIMAL(10,2), NOT NULL)

    StockQuantity (INT)

**Customers:**

    CustomerID (INT, PRIMARY KEY)

    CustomerName (VARCHAR(100), NOT NULL)

    Email (VARCHAR(100), UNIQUE)

    JoinDate (DATE)

**Orders:**

    OrderID (INT, PRIMARY KEY)

    CustomerID (INT, FOREIGN KEY → Customers)

    OrderDate (DATE, NOT NULL)

    TotalAmount (DECIMAL(10,2))

#**2. Insert the following records into each table**

**Categories**

|CategoryID|	Category Name|
|:--|:--|
|1|	Electronics|
|2 |Books|
|3|	Home Goods|
|4|	Apparel|

**Products**

|ProductID|	ProductName|	CategoryID|	Price|	StockQuantity|
|:--|:--|:--|:--|:--|
|101|	Laptop Pro|	1|	1200.00|	50|
|102|	SQL Handbook|	2	|45.50|	200|
|103|	Smart Speaker|	1|	99.99|	150|
|104|	Coffee Maker|	3|	75.00|	80|
|105|	Novel : The Great SQL|	2| 5.00|	120|
|106|	Wireless Earbuds	|1|	150.00|	100|
|107|Blender X|	3	|120.00|	60|
|108	|T-Shirt Casual|	4|	20.00|	300|

**Customers**

|CustomerID|	CustomerName|	Email|	Joining Date|
|:--|:--|:--|:--|
|1|	Alice Wonderland|	alice@example.com|2023-01-10|
|2|	Bob the Builder|	bob@example.com|2022-11-25|
|3|	Charlie Chaplin|	charlie@example.com|2023-03-01|
|4|	Diana Prince|	diana@example.com| 2021-04-26|

**Orders**

|OrderID|	CustomerID|	OrderDate|	TotalAmount|
|:--|:--|:--|:--|
|1001|	1	|2023-04-26|	1245.50|
|1002|	2	|2023-10-12|	99.99|
|1003	|1	|2023-07-01|	145.00|
|1004|	3	|2023-01-14|	150.00|
|1005	|2	|2023-09-24|	120.00|
|1006|	1	|2023-06-19|	20.00|

## Connect to Database

### Subtask:
Establish a connection to an in-memory SQLite database named `ECommerceDB`.


**Reasoning**:
The subtask requires connecting to an in-memory SQLite database and creating a cursor. This can be achieved by importing `sqlite3`, then establishing the connection using `sqlite3.connect(':memory:')`, and finally creating a cursor from the connection.



In [ ]:
import sqlite3

# Establish a connection to an in-memory SQLite database
conn_ecommerce = sqlite3.connect(':memory:')
cursor_ecommerce = conn_ecommerce.cursor()

print("Connected to an in-memory SQLite database for ECommerceDB.")

Connected to an in-memory SQLite database for ECommerceDB.


**Reasoning**:
The next step is to create the 'Categories' table as defined in the problem description, including its columns, data types, and constraints (PRIMARY KEY, NOT NULL, UNIQUE). This is a DDL operation.



In [ ]:
create_categories_table_sql = """
CREATE TABLE Categories (
    CategoryID INT PRIMARY KEY,
    CategoryName VARCHAR(50) NOT NULL UNIQUE
);
"""
cursor_ecommerce.execute(create_categories_table_sql)
conn_ecommerce.commit()
print("Table 'Categories' created successfully.")

Table 'Categories' created successfully.


**Reasoning**:
The next step is to create the 'Products' table as defined in the problem description, including its columns, data types, and constraints (PRIMARY KEY, NOT NULL, UNIQUE, and FOREIGN KEY). This is a DDL operation.



In [ ]:
create_products_table_sql = """
CREATE TABLE Products (
    ProductID INT PRIMARY KEY,
    ProductName VARCHAR(100) NOT NULL UNIQUE,
    CategoryID INT,
    Price DECIMAL(10,2) NOT NULL,
    StockQuantity INT,
    FOREIGN KEY (CategoryID) REFERENCES Categories(CategoryID)
);
"""
cursor_ecommerce.execute(create_products_table_sql)
conn_ecommerce.commit()
print("Table 'Products' created successfully.")

Table 'Products' created successfully.


**Reasoning**:
The next step is to create the 'Customers' table as defined in the problem description, including its columns, data types, and constraints (PRIMARY KEY, NOT NULL, UNIQUE, DATE). This is a DDL operation.



In [ ]:
create_customers_table_sql = """
CREATE TABLE Customers (
    CustomerID INT PRIMARY KEY,
    CustomerName VARCHAR(100) NOT NULL,
    Email VARCHAR(100) UNIQUE,
    JoinDate DATE
);
"""
cursor_ecommerce.execute(create_customers_table_sql)
conn_ecommerce.commit()
print("Table 'Customers' created successfully.")

Table 'Customers' created successfully.


**Reasoning**:
The next step is to create the 'Orders' table as defined in the problem description, including its columns, data types, and constraints (PRIMARY KEY, NOT NULL, and FOREIGN KEY). This is a DDL operation.



In [ ]:
create_orders_table_sql = """
CREATE TABLE Orders (
    OrderID INT PRIMARY KEY,
    CustomerID INT,
    OrderDate DATE NOT NULL,
    TotalAmount DECIMAL(10,2),
    FOREIGN KEY (CustomerID) REFERENCES Customers(CustomerID)
);
"""
cursor_ecommerce.execute(create_orders_table_sql)
conn_ecommerce.commit()
print("Table 'Orders' created successfully.")

Table 'Orders' created successfully.


**Reasoning**:
Now that all tables are created, the next step is to insert the provided sample data into the `Categories` table. This is a DML operation.



In [ ]:
insert_categories_data_sql = [
    (1, 'Electronics'),
    (2, 'Books'),
    (3, 'Home Goods'),
    (4, 'Apparel')
]
cursor_ecommerce.executemany("INSERT INTO Categories (CategoryID, CategoryName) VALUES (?, ?)", insert_categories_data_sql)
conn_ecommerce.commit()
print("Data inserted into 'Categories' table successfully.")

Data inserted into 'Categories' table successfully.


**Reasoning**:
Following the successful insertion of data into the `Categories` table, the next step is to insert the provided sample data into the `Products` table, which is a DML operation.



In [ ]:
insert_products_data_sql = [
    (101, 'Laptop Pro', 1, 1200.00, 50),
    (102, 'SQL Handbook', 2, 45.50, 200),
    (103, 'Smart Speaker', 1, 99.99, 150),
    (104, 'Coffee Maker', 3, 75.00, 80),
    (105, 'Novel : The Great SQL', 2, 5.00, 120),
    (106, 'Wireless Earbuds', 1, 150.00, 100),
    (107, 'Blender X', 3, 120.00, 60),
    (108, 'T-Shirt Casual', 4, 20.00, 300)
]
cursor_ecommerce.executemany(
    "INSERT INTO Products (ProductID, ProductName, CategoryID, Price, StockQuantity) VALUES (?, ?, ?, ?, ?)",
    insert_products_data_sql
)
conn_ecommerce.commit()
print("Data inserted into 'Products' table successfully.")

Data inserted into 'Products' table successfully.


**Reasoning**:
Now that data has been inserted into the `Products` table, the next step is to insert the provided sample data into the `Customers` table. This is a DML operation.



In [ ]:
insert_customers_data_sql = [
    (1, 'Alice Wonderland', 'alice@example.com', '2023-01-10'),
    (2, 'Bob the Builder', 'bob@example.com', '2022-11-25'),
    (3, 'Charlie Chaplin', 'charlie@example.com', '2023-03-01'),
    (4, 'Diana Prince', 'diana@example.com', '2021-04-26')
]
cursor_ecommerce.executemany(
    "INSERT INTO Customers (CustomerID, CustomerName, Email, JoinDate) VALUES (?, ?, ?, ?)",
    insert_customers_data_sql
)
conn_ecommerce.commit()
print("Data inserted into 'Customers' table successfully.")

Data inserted into 'Customers' table successfully.


**Reasoning**:
Now that data has been inserted into the `Customers` table, the next step is to insert the provided sample data into the `Orders` table, which is a DML operation.



In [ ]:
insert_orders_data_sql = [
    (1001, 1, '2023-04-26', 1245.50),
    (1002, 2, '2023-10-12', 99.99),
    (1003, 1, '2023-07-01', 145.00),
    (1004, 3, '2023-01-14', 150.00),
    (1005, 2, '2023-09-24', 120.00),
    (1006, 1, '2023-06-19', 20.00)
]
cursor_ecommerce.executemany(
    "INSERT INTO Orders (OrderID, CustomerID, OrderDate, TotalAmount) VALUES (?, ?, ?, ?)",
    insert_orders_data_sql
)
conn_ecommerce.commit()
print("Data inserted into 'Orders' table successfully.")

Data inserted into 'Orders' table successfully.


**Reasoning**:
Now that all data has been inserted into the `Orders` table, the next step is to verify the successful data insertion for all tables by retrieving and displaying all records from the `Categories` table. This is a DQL operation.



In [ ]:
print("\n--- Verifying data in Categories table ---")
cursor_ecommerce.execute("SELECT * FROM Categories")
category_records = cursor_ecommerce.fetchall()
for record in category_records:
    print(record)


--- Verifying data in Categories table ---
(1, 'Electronics')
(2, 'Books')
(3, 'Home Goods')
(4, 'Apparel')


**Reasoning**:
Following the verification of the `Categories` table, the next step is to verify the successful data insertion for the `Products` table. This is a DQL operation.



In [ ]:
print(
"\n--- Verifying data in Products table ---"
)
cursor_ecommerce.execute("SELECT * FROM Products")
product_records = cursor_ecommerce.fetchall()
for record in product_records:
    print(record)


--- Verifying data in Products table ---
(101, 'Laptop Pro', 1, 1200, 50)
(102, 'SQL Handbook', 2, 45.5, 200)
(103, 'Smart Speaker', 1, 99.99, 150)
(104, 'Coffee Maker', 3, 75, 80)
(105, 'Novel : The Great SQL', 2, 5, 120)
(106, 'Wireless Earbuds', 1, 150, 100)
(107, 'Blender X', 3, 120, 60)
(108, 'T-Shirt Casual', 4, 20, 300)


**Reasoning**:
Following the verification of the `Products` table, the next step is to verify the successful data insertion for the `Customers` table. This is a DQL operation.



In [ ]:
print(
"\n--- Verifying data in Customers table ---"
)
cursor_ecommerce.execute("SELECT * FROM Customers")
customer_records = cursor_ecommerce.fetchall()
for record in customer_records:
    print(record)


--- Verifying data in Customers table ---
(1, 'Alice Wonderland', 'alice@example.com', '2023-01-10')
(2, 'Bob the Builder', 'bob@example.com', '2022-11-25')
(3, 'Charlie Chaplin', 'charlie@example.com', '2023-03-01')
(4, 'Diana Prince', 'diana@example.com', '2021-04-26')


**Reasoning**:
Following the verification of the `Customers` table, the next step is to verify the successful data insertion for the `Orders` table. This is a DQL operation.



In [ ]:
print(
"\n--- Verifying data in Orders table ---"
)
cursor_ecommerce.execute("SELECT * FROM Orders")
order_records = cursor_ecommerce.fetchall()
for record in order_records:
    print(record)


--- Verifying data in Orders table ---
(1001, 1, '2023-04-26', 1245.5)
(1002, 2, '2023-10-12', 99.99)
(1003, 1, '2023-07-01', 145)
(1004, 3, '2023-01-14', 150)
(1005, 2, '2023-09-24', 120)
(1006, 1, '2023-06-19', 20)


**Reasoning**:
All tables have been created, data inserted, and verified. The final step is to close the database connection, as specified in the problem description.



In [ ]:
conn_ecommerce.close()
print("ECommerceDB database connection closed.")

ECommerceDB database connection closed.


## Summary:

### Q&A
The successful creation of the `ECommerceDB` database, all tables, and the insertion of specified data were confirmed, and the database connection was successfully closed.

### Data Analysis Key Findings
*   An in-memory SQLite database named `ECommerceDB` was successfully established.
*   Four tables - `Categories`, `Products`, `Customers`, and `Orders` - were successfully created, adhering to their specified schemas and constraints, including primary and foreign keys, unique, and not null constraints.
*   Data was successfully inserted into all tables:
    *   4 records into the `Categories` table (e.g., 'Electronics', 'Books').
    *   8 records into the `Products` table (e.g., 'Laptop Pro', 'SQL Handbook').
    *   4 records into the `Customers` table (e.g., 'Alice Wonderland', 'Bob the Builder').
    *   6 records into the `Orders` table.
*   Verification steps confirmed that all inserted data was correctly present in each respective table.
*   The database connection to `ECommerceDB` was successfully closed after all operations were completed.

### Insights or Next Steps
*   The successfully created and populated E-commerce database serves as a robust foundation for running various analytical queries to understand sales trends, customer behavior, or product performance.
*   Further analysis could involve writing SQL queries to perform joins between these tables to answer more complex business questions, such as identifying top-selling products by category or customers with the highest total purchase amounts.


#**Question 7 : Generate a report showing CustomerName, Email, and the TotalNumberofOrders for each customer.**

In [ ]:
report_query = """
SELECT
    c.FirstName || ' ' || c.LastName AS CustomerName,
    c.Email,
    COUNT(o.OrderID) AS TotalNumberOfOrders
FROM
    Customers c
LEFT JOIN
    Orders o ON c.CustomerID = o.CustomerID
GROUP BY
    c.CustomerID, c.FirstName, c.LastName, c.Email
ORDER BY
    CustomerName;
"""

cursor_ecommerce.execute(report_query)
report_results = cursor_ecommerce.fetchall()

print("Customer Order Report:")
print("----------------------")
for row in report_results:
    print(f"Name: {row[0]}, Email: {row[1]}, Total Orders: {row[2]}")

# It's good practice to close the connection when done, but for this exercise, we might keep it open if more queries are expected.
# conn_ecommerce.close()
# print("\nDatabase connection to ECommerceDB.db closed.")

Customer Order Report:
----------------------
Name: Alice Smith, Email: alice.smith@example.com, Total Orders: 2
Name: Bob Johnson, Email: bob.j@example.com, Total Orders: 1
Name: Charlie Brown, Email: charlie.b@example.com, Total Orders: 0


#**Question 8 : Retrieve Product Information with Category.**

In [ ]:
product_query = """
SELECT
    ProductID,
    ProductName,
    Category,
    Price
FROM
    Products;
"""

cursor_ecommerce.execute(product_query)
product_results = cursor_ecommerce.fetchall()

print("Product Information:")
print("--------------------")
for row in product_results:
    print(f"ID: {row[0]}, Name: {row[1]}, Category: {row[2]}, Price: {row[3]}")

# The connection is kept open as per previous instruction, but would typically be closed after all queries.
# conn_ecommerce.close()
# print("\nDatabase connection to ECommerceDB.db closed.")

Product Information:
--------------------
ID: 1, Name: Laptop, Category: Electronics, Price: 1200.0
ID: 2, Name: Mouse, Category: Electronics, Price: 25.5
ID: 3, Name: Keyboard, Category: Electronics, Price: 75.0
ID: 4, Name: Desk Chair, Category: Furniture, Price: 150.0


#**Question 9 : Display top 2 most expensive products in each category using CTE and Window Function.**

In [ ]:
top_products_query = """
WITH RankedProducts AS (
    SELECT
        ProductID,
        ProductName,
        Category,
        Price,
        ROW_NUMBER() OVER (PARTITION BY Category ORDER BY Price DESC) as rn
    FROM
        Products
)
SELECT
    ProductID,
    ProductName,
    Category,
    Price
FROM
    RankedProducts
WHERE
    rn <= 2
ORDER BY
    Category, Price DESC;
"""

cursor_ecommerce.execute(top_products_query)
top_products_results = cursor_ecommerce.fetchall()

print("Top 2 Most Expensive Products per Category:")
print("-------------------------------------------")
for row in top_products_results:
    print(f"Category: {row[2]}, Name: {row[1]}, Price: {row[3]}, ID: {row[0]}")

# The connection is kept open for further queries if needed.

Top 2 Most Expensive Products per Category:
-------------------------------------------
Category: Electronics, Name: Laptop, Price: 1200.0, ID: 1
Category: Electronics, Name: Keyboard, Price: 75.0, ID: 3
Category: Furniture, Name: Desk Chair, Price: 150.0, ID: 4


#**Question 10 : Sakila Database Business Analysis**


**You are hired as a data analyst by Sakila Video Rentals, a global movie rental company. The management team is looking to improve decision-making by analyzing existing customer, rental, and inventory data.**

**Using the Sakila database, answer the following business questions to support key strategic initiatives.**

#**Tasks & Questions:**

    Identify the top 5 customers based on the total amount they’ve spent. Include customer name, email, and total amount spent.

    Which 3 movie categories have the highest rental counts? Display the category name and number of times movies from that category were rented.

    Calculate how many films are available at each store and how many of those have never been rented.

    Show the total revenue per month for the year 2023 to analyze business seasonality.

    Identify customers who have rented more than 10 times in the last 6 months.

## Connect to Sakila-like Database

### Subtask:
Establish a connection to a new in-memory SQLite database to simulate the Sakila database for demonstration purposes.


**Reasoning**:
The subtask requires establishing a connection to an in-memory SQLite database and creating a cursor object, followed by a confirmation message.



In [ ]:
import sqlite3

# Establish a connection to an in-memory SQLite database
conn_sakila = sqlite3.connect(':memory:')
cursor_sakila = conn_sakila.cursor()

print("Connected to an in-memory SQLite database for Sakila simulation.")

Connected to an in-memory SQLite database for Sakila simulation.


**Reasoning**:
The next step is to create the `customer` table, as specified in the problem description, including appropriate columns and constraints, to simulate the Sakila database for the first part of Question 10.



In [ ]:
create_customer_table_sql = """
CREATE TABLE customer (
    customer_id INTEGER PRIMARY KEY AUTOINCREMENT,
    first_name VARCHAR(50) NOT NULL,
    last_name VARCHAR(50) NOT NULL,
    email VARCHAR(100) UNIQUE,
    create_date DATE NOT NULL
);
"""
cursor_sakila.execute(create_customer_table_sql)
conn_sakila.commit()
print("Table 'customer' created successfully.")

Table 'customer' created successfully.


**Reasoning**:
To answer the question about top customers by total amount spent, both `customer` and `payment` tables are required. Now that the `customer` table is created, the next step is to create the `payment` table with appropriate columns and constraints as defined in a typical Sakila schema.



In [ ]:
create_payment_table_sql = """
CREATE TABLE payment (
    payment_id INTEGER PRIMARY KEY AUTOINCREMENT,
    customer_id INTEGER NOT NULL,
    amount DECIMAL(5,2) NOT NULL,
    payment_date DATETIME NOT NULL,
    FOREIGN KEY (customer_id) REFERENCES customer(customer_id)
);
"""
cursor_sakila.execute(create_payment_table_sql)
conn_sakila.commit()
print("Table 'payment' created successfully.")

Table 'payment' created successfully.


**Reasoning**:
Now that both `customer` and `payment` tables are created, the next step is to insert sample data into the `customer` table to populate it, which is a DML operation required before inserting payments.



In [ ]:
insert_customer_data_sql = [
    (1, 'Alice', 'Smith', 'alice.smith@example.com', '2023-01-10'),
    (2, 'Bob', 'Johnson', 'bob.j@example.com', '2022-11-25'),
    (3, 'Charlie', 'Brown', 'charlie.b@example.com', '2023-03-01'),
    (4, 'Diana', 'Prince', 'diana.p@example.com', '2021-04-26'),
    (5, 'Eve', 'Adams', 'eve.a@example.com', '2023-05-15')
]
cursor_sakila.executemany(
    "INSERT INTO customer (customer_id, first_name, last_name, email, create_date) VALUES (?, ?, ?, ?, ?)",
    insert_customer_data_sql
)
conn_sakila.commit()
print("Sample data inserted into 'customer' table successfully.")

Sample data inserted into 'customer' table successfully.


**Reasoning**:
Now that the `customer` table is populated, the next step is to insert sample data into the `payment` table, linking payments to the customers. This DML operation is crucial for calculating total amounts spent by customers, which is the first part of Question 10.



In [ ]:
insert_payment_data_sql = [
    (1, 1, 50.00, '2023-01-15 10:00:00'),
    (2, 1, 25.50, '2023-02-01 11:30:00'),
    (3, 2, 30.00, '2022-12-01 14:00:00'),
    (4, 1, 15.00, '2023-03-10 09:00:00'),
    (5, 3, 40.00, '2023-03-05 16:00:00'),
    (6, 2, 10.00, '2023-04-20 18:00:00'),
    (7, 5, 60.00, '2023-05-20 12:00:00'),
    (8, 1, 5.00, '2023-06-01 13:00:00'),
    (9, 3, 20.00, '2023-07-11 10:00:00'),
    (10, 4, 35.00, '2022-05-01 17:00:00'),
    (11, 2, 5.00, '2023-08-01 09:00:00'),
    (12, 5, 20.00, '2023-09-15 14:00:00')
]
cursor_sakila.executemany(
    "INSERT INTO payment (payment_id, customer_id, amount, payment_date) VALUES (?, ?, ?, ?)",
    insert_payment_data_sql
)
conn_sakila.commit()
print("Sample data inserted into 'payment' table successfully.")

Sample data inserted into 'payment' table successfully.


**Reasoning**:
Now that both the `customer` and `payment` tables are populated, the next step is to execute a DQL query to identify the top 5 customers based on the total amount they have spent, as required by the first part of Question 10. This involves joining the two tables, grouping by customer, summing the payments, and ordering the results.



In [ ]:
top_customers_query = """
SELECT
    c.first_name,
    c.last_name,
    c.email,
    SUM(p.amount) AS total_amount_spent
FROM
    customer c
JOIN
    payment p ON c.customer_id = p.customer_id
GROUP BY
    c.customer_id, c.first_name, c.last_name, c.email
ORDER BY
    total_amount_spent DESC
LIMIT 5;
"""

cursor_sakila.execute(top_customers_query)
top_customers_results = cursor_sakila.fetchall()

print("\nTop 5 Customers by Total Amount Spent:")
print("---------------------------------------")
for row in top_customers_results:
    print(f"Name: {row[0]} {row[1]}, Email: {row[2]}, Total Spent: {row[3]}")


Top 5 Customers by Total Amount Spent:
---------------------------------------
Name: Alice Smith, Email: alice.smith@example.com, Total Spent: 95.5
Name: Eve Adams, Email: eve.a@example.com, Total Spent: 80
Name: Charlie Brown, Email: charlie.b@example.com, Total Spent: 60
Name: Bob Johnson, Email: bob.j@example.com, Total Spent: 45
Name: Diana Prince, Email: diana.p@example.com, Total Spent: 35


### Which 3 movie categories have the highest rental counts?

To identify the top 3 movie categories with the highest rental counts, we would need to join the `category`, `film_category`, `inventory`, and `rental` tables. The query would group the results by category name and count the rentals for each category.

**Assumed Sakila Schema Tables and Relationships:**
*   `category`: `category_id`, `name`
*   `film_category`: `film_id`, `category_id` (links `film` to `category`)
*   `inventory`: `inventory_id`, `film_id`, `store_id`
*   `rental`: `rental_id`, `inventory_id`

**SQL Query:**

```sql
SELECT
    c.name AS category_name,
    COUNT(r.rental_id) AS rental_count
FROM
    category c
JOIN
    film_category fc ON c.category_id = fc.category_id
JOIN
    inventory i ON fc.film_id = i.film_id
JOIN
    rental r ON i.inventory_id = r.inventory_id
GROUP BY
    c.name
ORDER BY
    rental_count DESC
LIMIT 3;
```

**Explanation:**
1.  **`FROM category c`**: Starts with the `category` table to get category names.
2.  **`JOIN film_category fc ON c.category_id = fc.category_id`**: Links categories to films.
3.  **`JOIN inventory i ON fc.film_id = i.film_id`**: Links films to inventory items, which represent physical copies available for rent.
4.  **`JOIN rental r ON i.inventory_id = r.inventory_id`**: Links inventory items to actual rental events.
5.  **`GROUP BY c.name`**: Groups the results by each category name.
6.  **`COUNT(r.rental_id)`**: Counts the number of rentals for each grouped category.
7.  **`ORDER BY rental_count DESC`**: Sorts the categories in descending order based on their rental count.
8.  **`LIMIT 3`**: Restricts the output to the top 3 categories.

## Explain Sub-Question 3: Films Available and Never Rented

### Subtask:
Provide the SQL query and explanation for calculating how many films are available at each store and how many have never been rented, assuming a full Sakila schema (film, inventory, rental, store).


## SQL Query for Films Available and Never Rented

```sql
SELECT
    s.store_id,
    COUNT(DISTINCT i.film_id) AS films_available_at_store,
    SUM(CASE WHEN r.rental_id IS NULL THEN 1 ELSE 0 END) AS films_never_rented
FROM
    store s
JOIN
    inventory i ON s.store_id = i.store_id
LEFT JOIN
    rental r ON i.inventory_id = r.inventory_id
GROUP BY
    s.store_id
ORDER BY
    s.store_id;
```


### Explanation of the SQL Query:

This SQL query is designed to calculate two metrics for each store: the total number of unique films available and the number of those films that have never been rented.

1.  **`FROM store s`**: We start by selecting from the `store` table, aliased as `s`, as we want to aggregate results per store.

2.  **`JOIN inventory i ON s.store_id = i.store_id`**: We join with the `inventory` table, aliased as `i`. The `inventory` table lists all the copies of films available in each store. This join links each store to its available film inventory.

3.  **`LEFT JOIN rental r ON i.inventory_id = r.inventory_id`**: A `LEFT JOIN` is used with the `rental` table, aliased as `r`. This is crucial because we want to include all inventory items (films) from each store, even if they have never been rented. If an `inventory_id` from the `inventory` table has no matching record in the `rental` table, the columns from the `rental` table will be `NULL` for that row. This allows us to identify unrented films.

4.  **`GROUP BY s.store_id`**: The results are grouped by `store_id` to perform aggregations (counts and sums) for each individual store.

5.  **`SELECT s.store_id, COUNT(DISTINCT i.film_id) AS films_available_at_store, SUM(CASE WHEN r.rental_id IS NULL THEN 1 ELSE 0 END) AS films_never_rented`**:
    *   **`s.store_id`**: Selects the ID of the store.
    *   **`COUNT(DISTINCT i.film_id) AS films_available_at_store`**: Counts the number of unique `film_id`s in the `inventory` for each store. This represents the total unique films stocked at that store.
    *   **`SUM(CASE WHEN r.rental_id IS NULL THEN 1 ELSE 0 END) AS films_never_rented`**: This is a conditional sum. For each row in the joined result, if `r.rental_id` is `NULL` (meaning the particular inventory item has never been rented), it counts as `1`. Otherwise, it counts as `0`. The `SUM` then tallies up all inventory items that have never been rented for each store. This accurately identifies films that have never had a rental record associated with them.

6.  **`ORDER BY s.store_id`**: Orders the final result set by store ID for better readability.

## Explain Sub-Question 4: Total Revenue Per Month for 2023

### Subtask:
Provide the SQL query and explanation for showing the total revenue per month for the year 2023, assuming a full Sakila schema (payment).

### Explanation of the SQL Query:

```sql
SELECT
    strftime('%Y-%m', payment_date) AS payment_month,
    SUM(amount) AS total_revenue
FROM
    payment
WHERE
    strftime('%Y', payment_date) = '2023'
GROUP BY
    payment_month
ORDER BY
    payment_month;
```

**1. `SELECT strftime('%Y-%m', payment_date) AS payment_month, SUM(amount) AS total_revenue`**
   - `strftime('%Y-%m', payment_date)`: This SQL function extracts the year and month from the `payment_date` column and formats it as 'YYYY-MM' (e.g., '2023-01'). This formatted string will represent each month.
   - `AS payment_month`: This assigns an alias `payment_month` to the extracted year-month string, making the output column more readable.
   - `SUM(amount)`: This aggregate function calculates the sum of all `amount` values for each group (which will be each month).
   - `AS total_revenue`: This assigns an alias `total_revenue` to the sum of amounts.

**2. `FROM payment`**
   - This specifies that the query will retrieve data from the `payment` table.

**3. `WHERE strftime('%Y', payment_date) = '2023'`**
   - `strftime('%Y', payment_date)`: This extracts only the year from the `payment_date` column.
   - `= '2023'`: This condition filters the records to include only those payments made in the year 2023. This is crucial for isolating the data to the requested year.

**4. `GROUP BY payment_month`**
   - This clause groups the rows based on the `payment_month` (the 'YYYY-MM' string). All payments occurring in the same month will be grouped together, allowing the `SUM(amount)` function to calculate the total revenue for that specific month.

**5. `ORDER BY payment_month`**
   - This clause sorts the final result set in ascending order by `payment_month`, ensuring that the monthly revenues are displayed chronologically from January to December.

**Reasoning**:
The explanation for the SQL query has been provided in the previous step. Now, I need to execute the SQL query to demonstrate the result of showing the total revenue per month for the year 2023, using the in-memory Sakila database created previously.



In [ ]:
total_revenue_2023_query = """
SELECT
    strftime('%Y-%m', payment_date) AS payment_month,
    SUM(amount) AS total_revenue
FROM
    payment
WHERE
    strftime('%Y', payment_date) = '2023'
GROUP BY
    payment_month
ORDER BY
    payment_month;
"""

cursor_sakila.execute(total_revenue_2023_query)
monthly_revenue_results = cursor_sakila.fetchall()

print("\nTotal Revenue Per Month for 2023:")
print("--------------------------------")
if not monthly_revenue_results:
    print("No revenue data for 2023 in the sample dataset.")
for row in monthly_revenue_results:
    print(f"Month: {row[0]}, Total Revenue: {row[1]:.2f}")


Total Revenue Per Month for 2023:
--------------------------------
Month: 2023-01, Total Revenue: 50.00
Month: 2023-02, Total Revenue: 25.50
Month: 2023-03, Total Revenue: 55.00
Month: 2023-04, Total Revenue: 10.00
Month: 2023-05, Total Revenue: 60.00
Month: 2023-06, Total Revenue: 5.00
Month: 2023-07, Total Revenue: 20.00
Month: 2023-08, Total Revenue: 5.00
Month: 2023-09, Total Revenue: 20.00


## Explain Sub-Question 5: Customers with More Than 10 Rentals in the Last 6 Months

### Subtask:
Provide the SQL query and explanation for identifying customers who have rented more than 10 times in the last 6 months, assuming a full Sakila schema (customer, rental).

### Assumed Sakila Schema Tables and Relationships:
*   `customer`: `customer_id`, `first_name`, `last_name`, `email`
*   `rental`: `rental_id`, `customer_id`, `rental_date`

### SQL Query:

```sql
SELECT
    c.first_name,
    c.last_name,
    c.email,
    COUNT(r.rental_id) AS rental_count
FROM
    customer c
JOIN
    rental r ON c.customer_id = r.customer_id
WHERE
    r.rental_date >= date('now', '-6 months') -- Filters rentals from the last 6 months
GROUP BY
    c.customer_id, c.first_name, c.last_name, c.email
HAVING
    COUNT(r.rental_id) > 10
ORDER BY
    rental_count DESC;
```

### Explanation of the SQL Query:

This SQL query identifies customers who have made more than 10 rentals within the last six months.

1.  **`FROM customer c`**: We start by selecting from the `customer` table, aliased as `c`, to get customer details.

2.  **`JOIN rental r ON c.customer_id = r.customer_id`**: We join the `customer` table with the `rental` table, aliased as `r`, on `customer_id`. This links each customer to their rental records.

3.  **`WHERE r.rental_date >= date('now', '-6 months')`**: This `WHERE` clause filters the rental records to include only those that occurred within the last six months from the current date.
    *   `date('now')` gets the current date.
    *   `'-6 months'` subtracts six months from the current date.

4.  **`GROUP BY c.customer_id, c.first_name, c.last_name, c.email`**: The results are grouped by customer to count the total rentals for each individual customer.

5.  **`SELECT c.first_name, c.last_name, c.email, COUNT(r.rental_id) AS rental_count`**:
    *   Selects the customer's first name, last name, and email.
    *   `COUNT(r.rental_id)` calculates the total number of rentals for each customer within the filtered time period.
    *   `AS rental_count` assigns an alias to the count column for readability.

6.  **`HAVING COUNT(r.rental_id) > 10`**: This `HAVING` clause filters the grouped results, keeping only those customers who have a `rental_count` greater than 10. This is crucial for fulfilling the condition of "more than 10 times."

7.  **`ORDER BY rental_count DESC`**: Finally, the results are ordered in descending order based on the `rental_count`, so customers with the highest number of rentals appear first.

## Explain Task 5: Customers with More Than 10 Rentals in Last 6 Months





## SQL Query for Customers with More Than 10 Rentals in Last 6 Months

```sql
SELECT
    c.first_name,
    c.last_name,
    c.email,
    COUNT(r.rental_id) AS total_rentals
FROM
    customer c
JOIN
    rental r ON c.customer_id = r.customer_id
WHERE
    r.rental_date >= date('now', '-6 months')
GROUP BY
    c.customer_id, c.first_name, c.last_name, c.email
HAVING
    COUNT(r.rental_id) > 10
ORDER BY
    total_rentals DESC;
```

### Explanation of the SQL Query:

This SQL query is designed to identify customers who have made more than 10 rentals within the last six months.

1.  **`SELECT c.first_name, c.last_name, c.email, COUNT(r.rental_id) AS total_rentals`**
    *   `c.first_name`, `c.last_name`, `c.email`: These select the customer's first name, last name, and email address to identify them in the report.
    *   `COUNT(r.rental_id) AS total_rentals`: This counts the number of rental records (`rental_id`) for each customer. The result is aliased as `total_rentals` for clarity.

2.  **`FROM customer c JOIN rental r ON c.customer_id = r.customer_id`**
    *   `FROM customer c`: The query starts by selecting from the `customer` table, aliased as `c`.
    *   `JOIN rental r ON c.customer_id = r.customer_id`: This joins the `customer` table with the `rental` table (aliased as `r`) using `customer_id`. This linkage is essential to associate rentals with specific customers.

3.  **`WHERE r.rental_date >= date('now', '-6 months')`**
    *   `WHERE r.rental_date >= date('now', '-6 months')`: This is the filtering condition. It restricts the rental records to only those that occurred within the last six months from the current date. `date('now', '-6 months')` is a SQLite function that calculates the date exactly six months ago.

4.  **`GROUP BY c.customer_id, c.first_name, c.last_name, c.email`**
    *   `GROUP BY`: This clause groups the rows by each unique customer. All aggregate functions (like `COUNT()`) will then operate on these groups. Including `first_name`, `last_name`, and `email` in the `GROUP BY` clause is standard practice when these columns are selected without being part of an aggregate function.

5.  **`HAVING COUNT(r.rental_id) > 10`**
    *   `HAVING`: This clause filters the *grouped* results. It ensures that only those customer groups (i.e., individual customers) whose `total_rentals` count is greater than 10 are included in the final output.

6.  **`ORDER BY total_rentals DESC`**
    *   `ORDER BY total_rentals DESC`: This sorts the final result set in descending order based on the `total_rentals` count, showing the customers with the highest number of rentals at the top.

## Close Sakila-like Database Connection

### Subtask:
Close the connection to the in-memory 'Sakila-like' database.


## Summary:

### Q&A

1.  **Who are the top 5 customers by total amount spent in the simulated database?**
    The top 5 customers by total amount spent are:
    *   Alice Smith: \$95.50
    *   Eve Adams: \$80.00
    *   Charlie Brown: \$60.00
    *   Bob Johnson: \$45.00
    *   Diana Prince: \$35.00

2.  **Which 3 movie categories have the highest rental counts (based on a full Sakila schema)?**
    The SQL query provided to identify the top 3 movie categories by rental counts involves joining the `category`, `film_category`, `inventory`, and `rental` tables, grouping by category name, and counting rental IDs, then ordering by rental count in descending order and limiting to 3. (Specific category names and counts cannot be provided without a full Sakila database execution.)

3.  **How many films are available at each store and how many have never been rented (based on a full Sakila schema)?**
    The SQL query provided calculates these metrics by joining `store` and `inventory` tables, then left joining `rental` to identify unrented items (where `rental_id` is NULL). It groups by `store_id`, counts `DISTINCT film_id` for available films, and sums a conditional count for never-rented films. (Specific counts per store cannot be provided without a full Sakila database execution.)

4.  **What is the total revenue per month for the year 2023 in the simulated database?**
    The total revenue per month for 2023 from the simulated data is:
    *   Month: 2023-01, Total Revenue: \$50.00
    *   Month: 2023-02, Total Revenue: \$25.50
    *   Month: 2023-03, Total Revenue: \$55.00
    *   Month: 2023-04, Total Revenue: \$10.00
    *   Month: 2023-05, Total Revenue: \$60.00
    *   Month: 2023-06, Total Revenue: \$5.00
    *   Month: 2023-07, Total Revenue: \$20.00
    *   Month: 2023-08, Total Revenue: \$5.00
    *   Month: 2023-09, Total Revenue: \$20.00

5.  **Which customers have rented more than 10 times in the last 6 months (based on a full Sakila schema)?**
    The SQL query provided identifies these customers by joining `customer` and `rental` tables, filtering rentals within the last 6 months using `date('now', '-6 months')`, grouping by customer, and using a `HAVING` clause to select customers with `COUNT(rental_id) > 10`. (Specific customer names and rental counts cannot be provided without a full Sakila database execution.)

### Data Analysis Key Findings

*   An in-memory SQLite database was successfully connected, and simplified `customer` and `payment` tables were created and populated with sample data.
*   **Customer Spending Analysis**: From the sample data, "Alice Smith" is the top spender with a total of \$95.50, followed by "Eve Adams" with \$80.00.
*   **Monthly Revenue Trend (2023)**: Based on the sample data, monthly revenue in 2023 peaked at \$60.00 in May and was lowest at \$5.00 in June and August.
*   **SQL Query Design for Sakila Schema**: Comprehensive SQL queries and detailed explanations were provided for several complex analytical tasks assuming a full Sakila schema, including:
    *   Identifying top movie categories by rental count using multiple joins and aggregation.
    *   Determining films available and never rented per store, utilizing `LEFT JOIN` to capture non-existent rental records.
    *   Calculating monthly revenue for a specific year (`2023`) by extracting year-month components from dates.
    *   Identifying high-frequency renters (more than 10 rentals in the last 6 months) using date functions, grouping, and a `HAVING` clause.


